src/iso.c

```diff
@@ -0,0 +1,535 @@
+    if (iso_media_peek_byte(iso_p) == 0) {
+        //No more records on this extent, but are there more on next extent?
+        if (iso_p->cur_extent < final_extent_of_directory) {
+            iso_media_seek(iso_p, (iso_p->cur_extent + 1) * ISO_BLOCKSIZE, SEEK_SET);
+            read_from->cur_offset = ISO_BLOCKSIZE * (iso_p->cur_extent - read_from->extent_offset);
+        }
+        else {
+            //End of this directory, return null
+            iso_media_close(iso_p);
+            return 0;
+        }
+    }
+
+    struct directory_record *next_dr = kmalloc(sizeof(*next_dr));
+
+    //Not at the end of the directory record, fill the dest by dereferencing dest_ptr
+    int success = get_directory_record(iso_p, next_dr);
+    if (!success) {
+        iso_media_close(iso_p);
+        return 0;
+    }
+
+    read_from->cur_offset = iso_p->cur_offset + (iso_p->cur_extent - read_from->extent_offset) * ISO_BLOCKSIZE;
+    //no need to reset extent, that shouldn't be an issue
+    iso_media_close(iso_p);
+    return next_dr;
+}
+
+//Documentation in header
+int iso_fclose(struct iso_file *file) {
+    if (file) {
+        kfree(file);
+        return 0;
+    }
+    return -1;
+}
+
+//Documentation in header
+struct iso_file *iso_fopen(const char *pname, int ata_unit) {
+    struct iso_file *file = kmalloc(sizeof(struct iso_file));
+    strcpy(file->pname, pname);
+    uint32_t dl;  //the data_length of the last item on the path
+    int extent_num = iso_look_up(pname, &dl, ata_unit, 0);
+    if (extent_num < 0) {
+        return 0;
+    }
+    file->cur_offset = 0;
+    file->extent_offset = extent_num;
+    file->ata_unit = ata_unit;
+    file->at_EOF = 0;
+    file->data_length = dl;
+
+    return file;
+}
+
+//Documentation in header
+int iso_fread(void *dest, int elem_size, int num_elem, struct iso_file *file) {
+    struct iso_point *iso_p = iso_media_open(file->ata_unit);
+    iso_media_seek(iso_p, ISO_BLOCKSIZE * file->extent_offset + file->cur_offset, SEEK_SET);
+
+    int bytes_to_disk_read = elem_size * num_elem;
+    int bytes_to_file_read = bytes_to_disk_read;
+    if ((bytes_to_file_read + file->cur_offset) > file->data_length) {
+        // Don't actually want to read off the end of the file
+        bytes_to_file_read = file->data_length - file->cur_offset + 1;
+        bytes_to_disk_read = bytes_to_file_read - 1;
+        file->at_EOF = 1;
+    }
+    int bytes_read = iso_media_read(dest, 1, bytes_to_disk_read, iso_p);
+    if (bytes_read != bytes_to_disk_read) {
+        iso_media_close(iso_p);
+        return -1;
+    }
+
+    iso_media_close(iso_p);
+    file->cur_offset += bytes_to_file_read;
+    return bytes_read;
+}
+
+/**
+ * @brief Acts as an fclose() for iso disk
+ * @details Frees the iso_point that was used to navigate the ISO disk.
+ *
+ * @param iso_p A pointer to the iso_point to close.
+ */
+static void iso_media_close(struct iso_point *iso_p) {
+    kfree(iso_p);
```

Comment:
No, because iso_media close doesn't return anything. iso_fclose returns a status so that we match the POSIX standard. For iso_media_close, it's internal, so the user doesn't care about the status. Either we freed it or it was never properly allocated, but either way, it's gone.